spmd/compiler/fusion.py

```diff
@@ -0,0 +1,182 @@
+    OP,
+    get_node_tensor_numel,
+    get_output_node,
+    graph_cleanup,
+    pretty_print_graph,
+)
+from .log_utils import rank0_debug
+
+logger: logging.Logger = logging.getLogger(__name__)
+
+# enum for the supported fusion comm types
+class CommType(str, Enum):
+    allreduce = "allreduce_"
+    allgather = "allgather_"
+    broadcast = "broadcast_"
+    reducescatter = "reduce_scatter_"
+    scatter = "scatter_"
+
+
+@dataclass
+class FusionElement:
+    """This class tracks the nodes for a DTensor expanded communication collective in the graph"""
+
+    in_graph: bool = False
+    comm_type: Optional[CommType] = None
+    node_list: Optional[List[fx.Node]] = field(default_factory=lambda: [])  # type: ignore
+    size: int = 0
+    prev_node: Optional[fx.Node] = None  # node that was before start of section
+    next_node: Optional[fx.Node] = None  # node that was after end of section
+    processed: bool = False
+    output_name: str = ""
+    clone_node: Optional[fx.Node] = None
+    comm_node: Optional[fx.Node] = None
+    wait_node: Optional[fx.Node] = None
+
+
+@dataclass
+class GraphInfo:
+    len: int = 0
+    global_buffer: Optional[fx.Node] = None
+    global_buffer_size: int = 0
+    output: Optional[fx.Node] = None
+    first: Optional[fx.Node] = None
+
+    def update_info(self, gm: fx.GraphModule) -> None:
+        """get the len, input and output nodes"""
+        graph_len = gm.graph._len
+        if not graph_len:
+            raise ValueError("Empty graph passed in....")
+        self.len = graph_len
+
+        nodelist = gm.graph.nodes
+
+        for i, node in enumerate(nodelist):
+            if node.op == OP.PLACEHOLDER:
+                self.first = node
+                break
+
+        self.output = get_output_node(gm)
+        assert (
+            self.output is not None
+        ), f"unable to locate output node in gm {gm.graph}"
+
+        rank0_debug(
+            logger,
+            f"updated graph_info - len = {self.len} input = {self.first}, output = {self.output}",
+        )
+
+
+def _insert_fusion_buffer_node(
+    gm: fx.GraphModule, insert_before_node: fx.Node, buffer_size: Iterable[int]
+) -> fx.Node:
+    """insert a torch.empty node in front of insert_before_node"""
+    with gm.graph.inserting_before(insert_before_node):
+        new_buffer_node = gm.graph.create_node(
+            OP.CALL_FUNCTION,
+            target=torch.empty,
+            args=tuple(buffer_size),  # TODO - need device
+        )
+    assert (
+        new_buffer_node is not None
+    ), f"failed to create buffer node, size={buffer_size}"
+
+    return new_buffer_node
+
+
+def _scan_graph_for_fusion_elements(
+    gm: fx.GraphModule,
+    comm_type: CommType = CommType.allreduce,
+) -> Optional[List[FusionElement]]:
+    """scan entire graph for matching sections of CommTensor style expansions
+    returns list of FusionElements that match CommType"""
+
+    element_list = []
+
+    fe_sequence = [
```

Comment:
It looks like this particular sequence order is important. Is that right? Or is it sufficient that the wait_comm is the last node that we look for. 

I am assuming this is to support the current allreduce fusion for DDP only? or does it support other variations?

Reply:
I believe it is important that we match the entire sequence, rather than just the wait_comm node.  
The reason is we know from the entire sequence that this is the dummy_add insertion/expansion, as opposed to a possible future insertion of only a wait_comm for other reasons. 
Longer term we may be able to revert back to the original approach of simply doing this work during the original DTensor expansion pass, but for now we are doing post expansion.  
Assuming that other collectives are inserted using the same dummy_add approach, then this sequence match will also work for them as well. The actual comm node is dynamic in the sequence, and setup to support all the ones that comm_tensor supports.